File: apps/server/src/api/links.rs
use serde::Serialize;

#[derive(Debug, Clone, Serialize)]
pub(super) struct Link {
    href: String,
}

impl Link {
    pub(super) fn new(href: impl Into<String>) -> Self {
        Self { href: href.into() }
    }
}

pub(super) fn api_href() -> String {
    "/api".to_string()
}

pub(super) fn health_href() -> String {
    "/health".to_string()
}

pub(super) fn user_href(user_id: &str) -> String {
    format!("/api/users/{user_id}")
}

pub(super) fn user_workspaces_href(user_id: &str) -> String {
    format!("/api/users/{user_id}/workspaces")
}

pub(super) fn user_workspaces_page_href(user_id: &str, page: u32, page_size: u32) -> String {
    format!(
        "{}?page={page}&pageSize={page_size}",
        user_workspaces_href(user_id)
    )
}

pub(super) fn workspace_href(user_id: &str, workspace_id: &str) -> String {
    format!("/api/users/{user_id}/workspaces/{workspace_id}")
}

pub(super) fn workspace_members_href(user_id: &str, workspace_id: &str) -> String {
    format!("{}/members", workspace_href(user_id, workspace_id))
}

pub(super) fn workspace_diagrams_href(workspace_id: &str) -> String {
    format!("/api/workspaces/{workspace_id}/diagrams")
}

pub(super) fn workspace_logical_entities_href(workspace_id: &str) -> String {
    format!("/api/workspaces/{workspace_id}/logical-entities")
}

pub(super) fn workspace_member_href(user_id: &str, workspace_id: &str, member_id: &str) -> String {
    format!(
        "{}/{}",
        workspace_members_href(user_id, workspace_id),
        member_id
    )
}

File: apps/server-nest/src/api/links.ts
export interface Link {
  href: string;
}

export function link(href: string): Link {
  return { href };
}

export function apiHref(): string {
  return '/api';
}

export function healthHref(): string {
  return '/health';
}

export function userHref(userId: string): string {
  return `/api/users/${userId}`;
}

export function userWorkspacesHref(userId: string): string {
  return `/api/users/${userId}/workspaces`;
}

export function userSidebarHref(userId: string): string {
  return `/api/users/${userId}/sidebar`;
}

export function userWorkspacesPageHref(
  userId: string,
  page: number,
  pageSize: number,
): string {
  return `${userWorkspacesHref(userId)}?page=${page}&pageSize=${pageSize}`;
}

export function workspaceHref(userId: string, workspaceId: string): string {
  return `/api/users/${userId}/workspaces/${workspaceId}`;
}

export function workspaceMembersHref(
  userId: string,
  workspaceId: string,
): string {
  return `${workspaceHref(userId, workspaceId)}/members`;
}

export function workspaceDiagramsHref(workspaceId: string): string {
  return `/api/workspaces/${workspaceId}/diagrams`;
}

export function workspaceLogicalEntitiesHref(workspaceId: string): string {
  return `/api/workspaces/${workspaceId}/logical-entities`;
}

export function workspaceLogicalRelationshipsHref(workspaceId: string): string {
  return `/api/workspaces/${workspaceId}/logical-relationships`;
}

export function workspaceMemberHref(
  userId: string,
  workspaceId: string,
  memberId: string,
): string {
  return `${workspaceMembersHref(userId, workspaceId)}/${memberId}`;
}
